src/pymor/algorithms/tr.py

```diff
@@ -0,0 +1,487 @@
+
+    See :func:`trust_region`.
+    """
+    if primal_dual:
+        surrogate = PrimalDualTRSurrogate(reductor, initial_guess)
+    else:
+        output = reductor.fom.output_functional
+        if isinstance(output, LincombOperator) and len(output.operators) == 1 and \
+            isinstance(output.operators[0], QuadraticFunctional):
+            # case for quadratic outputs. Should be obsolete soon, see QuadraticOutputTRSurrogate.
+            continuity_estimator_output = output.coefficients[0]
+            surrogate = QuadraticOutputTRSurrogate(reductor, initial_guess,
+                                                   continuity_estimator_output=continuity_estimator_output)
+        else:
+            surrogate = BasicTRSurrogate(reductor, initial_guess)
+
+    mu, data = trust_region(reductor.fom, surrogate, parameter_space=parameter_space, initial_guess=initial_guess,
+                            beta=beta, radius=radius, shrink_factor=shrink_factor, miniter=miniter,
+                            maxiter=maxiter, miniter_subproblem=miniter_subproblem,
+                            maxiter_subproblem=maxiter_subproblem, tol_criticality=tol_criticality,
+                            radius_tol=radius_tol, rtol_output=rtol_output, rtol_mu=rtol_mu, tol_sub=tol_sub,
+                            armijo_alpha=armijo_alpha, line_search_params=line_search_params,
+                            stagnation_window=stagnation_window, stagnation_threshold=stagnation_threshold)
+
+    # Note: this evaluation count assumes caching of the primal solution, see above.
+    if primal_dual:
+        fom_evaluations_outer = 0
+    else:
+        # for every gradient, one also requires the dual solution
+        fom_evaluations_outer = data['iterations']
+
+    data['fom_evaluations'] = surrogate.fom_evaluations + fom_evaluations_outer
+
+    return mu, data
+
+class TRSurrogate(BasicObject):
+    """Base class for :func:`trust_region` surrogates.
+
+    Not to be used directly.
+    """
+
+    def __init__(self, reductor, initial_guess, name='TRSurrogate'):
+        self.__auto_init(locals())
+        self.parameters = reductor.fom.parameters
+        self.dim_output = reductor.fom.dim_output
+
+        self.fom_evaluations = 0
+        self.rom_evaluations = 0
+        self.enrichments = 0
+
+        # generate a first rom based on the initial guess
+        if isinstance(initial_guess, Mu):
+            initial_guess = initial_guess.to_numpy()
+        assert isinstance(initial_guess, np.ndarray)
+        self.extend(initial_guess)
+        self.accept()
+
+        # so far, we only support 1-dimensional outputs
+        assert self.rom.dim_output == 1
+
+        # initialize placeholders for extension
+        self.new_reductor = None
+        self.new_rom = None
+
+    def output(self, mu):
+        self.rom_evaluations += 1
+        return self.rom.output(mu)
+
+    def output_d_mu(self, mu):
+        self.rom_evaluations += 2
```

Comment:
I guess this is only correct if the adjoint approach is used?

Reply:
Maybe it would make more sense to count the derivative evaluations separately?

Reply:
I guess the best way would be to have an `EvaluationCounter`-object for all `Models` but this is something for later.